src/gvmd.c

```diff
@@ -1619,25 +1651,163 @@ fork_agents_sync ()
 }
 #endif //ENABLE_AGENTS
 
+/**
+ * @brief Periodic timestamps for background jobs in the main loop.
+ */
+typedef struct {
+  time_t last_schedule;      ///< Last time the scheduler management was executed.
+  time_t last_feed_sync;     ///< Last time the feed sync was executed.
+  time_t last_queue;         ///< Last time queued task actions were executed.
+  time_t last_agents_sync;   ///< Last time the agents sync was executed.
+} periodic_times_t;
+
+/**
+ * @brief Return non-zero if @p period seconds have elapsed since @p last.
+ *
+ * @param[in] last    Previous run time (epoch seconds).
+ * @param[in] period  Minimum period in seconds.
+ * @param[in] now     Current time (epoch seconds).
+ *
+ * @return Non-zero if due, 0 otherwise.
+ */
+static int
+time_to_run (time_t last, time_t period, time_t now)
+{
+  return (now - last) >= period;
+}
+
+/**
+ * @brief Set a last-run timestamp to @p now.
+ *
+ * @param[out] last  Pointer to last-run timestamp to update.
+ * @param[in]  now   Current time (epoch seconds).
+ */
+static void
+set_now (time_t *last, time_t now)
+{
+  *last = now;
+}
+
+/**
+ * @brief Run scheduler management if due; updates last timestamp only on work.
+ *
+ * @param[in,out] t   Periodic timestamps; updates t->last_schedule on success.
+ * @param[in]     now Current time (epoch seconds).
+ *
+ * @return void. On fatal error, exits the process.
+ */
+static void
+run_schedule (periodic_times_t *t, time_t now)
+{
+  if (!time_to_run (t->last_schedule, SCHEDULE_PERIOD, now))
+    return;
+
+  int rc = manage_schedule (fork_connection_for_scheduler,
+                            scheduling_enabled,
+                            sigmask_normal);
+  switch (rc)
+    {
+    case 0:
+      set_now (&t->last_schedule, now);
```

Comment:
There's a slight difference with the previous behaviour here, because before it would use the time after the `manage_schedule`. Maybe this will have an effect if the schedule takes a long time?